File: lms/backend/routes/books.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from .. import schemas
from ..database import get_db
from ..models import Book
from ..utils import require_role

router = APIRouter(prefix="/books", tags=["books"])


@router.get("/", response_model=list[schemas.BookOut])
def list_books(q: str | None = Query(default=None), db: Session = Depends(get_db)):
    query = db.query(Book)
    if q:
        like = f"%{q}%"
        query = query.filter((Book.title.ilike(like)) | (Book.author.ilike(like)) | (Book.isbn.ilike(like)))
    return query.order_by(Book.created_at.desc()).all()


@router.post("/add", response_model=schemas.BookOut, dependencies=[Depends(require_role("librarian"))])
def add_book(book_in: schemas.BookCreate, db: Session = Depends(get_db)):
    exists = db.query(Book).filter(Book.isbn == book_in.isbn).first()
    if exists:
        raise HTTPException(status_code=400, detail="Book with this ISBN already exists")
    book = Book(
        title=book_in.title,
        author=book_in.author,
        isbn=book_in.isbn,
        total_copies=book_in.total_copies,
        available_copies=book_in.total_copies,
    )
    db.add(book)
    db.commit()
    db.refresh(book)
    return book


@router.put("/{book_id}", response_model=schemas.BookOut, dependencies=[Depends(require_role("librarian"))])
def update_book(book_id: int, book_in: schemas.BookUpdate, db: Session = Depends(get_db)):
    book = db.query(Book).filter(Book.id == book_id).first()
    if not book:
        raise HTTPException(status_code=404, detail="Book not found")
    if book_in.title is not None:
        book.title = book_in.title
    if book_in.author is not None:
        book.author = book_in.author
    if book_in.isbn is not None:
        # ensure uniqueness
        exists = db.query(Book).filter(Book.isbn == book_in.isbn, Book.id != book_id).first()
        if exists:
            raise HTTPException(status_code=400, detail="ISBN already used by another book")
        book.isbn = book_in.isbn
    if book_in.total_copies is not None:
        diff = book_in.total_copies - book.total_copies
        # adjust available based on new total but never below 0
        book.total_copies = book_in.total_copies
        book.available_copies = max(0, book.available_copies + diff)
    db.commit()
    db.refresh(book)
    return book


@router.delete("/{book_id}", dependencies=[Depends(require_role("librarian"))])
def delete_book(book_id: int, db: Session = Depends(get_db)):
    book = db.query(Book).filter(Book.id == book_id).first()
    if not book:
        raise HTTPException(status_code=404, detail="Book not found")
    if book.available_copies != book.total_copies:
        raise HTTPException(status_code=400, detail="Cannot delete: some copies are issued")
    db.delete(book)
    db.commit()
    return {"detail": "Deleted"}

File: lms/backend/routes/books.js
import express from 'express';
import db from '../db.js';
import { requireAuth } from '../middleware/auth.js';

const router = express.Router();

// Get all books with pagination and search
router.get('/', async (req, res) => {
  try {
    const page = parseInt(req.query.page) || 1;
    const limit = parseInt(req.query.limit) || 10;
    const offset = (page - 1) * limit;
    const search = req.query.q || '';
    const filter = req.query.filter || 'all'; // title, author, subject, all
    const subject = req.query.subject || '';
    const year = req.query.year || '';
    const sort = req.query.sort || 'title'; // title, author, year, popularity
    const order = req.query.order || 'ASC'; // ASC, DESC

    let whereConditions = [];
    let params = [];

    if (search) {
      if (filter === 'title') {
        whereConditions.push('title LIKE ?');
        params.push(`%${search}%`);
      } else if (filter === 'author') {
        whereConditions.push('author LIKE ?');
        params.push(`%${search}%`);
      } else if (filter === 'subject') {
        whereConditions.push('subject LIKE ?');
        params.push(`%${search}%`);
      } else {
        whereConditions.push('(title LIKE ? OR author LIKE ? OR isbn LIKE ? OR subject LIKE ?)');
        params.push(`%${search}%`, `%${search}%`, `%${search}%`, `%${search}%`);
      }
    }

    if (subject) {
      whereConditions.push('subject = ?');
      params.push(subject);
    }

    if (year) {
      whereConditions.push('publication_year = ?');
      params.push(year);
    }

    const whereClause = whereConditions.length > 0 
      ? `WHERE ${whereConditions.join(' AND ')}` 
      : '';

    let orderClause = '';
    if (sort === 'title') {
      orderClause = `ORDER BY title ${order}`;
    } else if (sort === 'author') {
      orderClause = `ORDER BY author ${order}`;
    } else if (sort === 'year') {
      orderClause = `ORDER BY publication_year ${order}`;
    } else if (sort === 'popularity') {
      orderClause = `ORDER BY (SELECT COUNT(*) FROM transactions WHERE transactions.book_id = books.id) ${order}`;
    } else {
      orderClause = 'ORDER BY title ASC';
    }

    const countSql = `SELECT COUNT(*) as total FROM books ${whereClause}`;
    const countResult = await db.get(countSql, params);
    const total = countResult.total;

    const sql = `
      SELECT books.*, 
             (SELECT COUNT(*) FROM transactions WHERE transactions.book_id = books.id) as issue_count
      FROM books 
      ${whereClause}
      ${orderClause}
      LIMIT ? OFFSET ?
    `;
    params.push(limit, offset);

    const books = await db.all(sql, params);

    res.json({
      books,
      pagination: {
        page,
        limit,
        total,
        pages: Math.ceil(total / limit)
      }
    });
  } catch (error) {
    console.error('Error fetching books:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Get book by ID
router.get('/:id', async (req, res) => {
  try {
    const book = await db.get('SELECT * FROM books WHERE id = ?', [req.params.id]);
    
    if (!book) {
      return res.status(404).json({ error: 'Book not found' });
    }

    // Get issue count
    const issueCount = await db.get(
      'SELECT COUNT(*) as count FROM transactions WHERE book_id = ?',
      [req.params.id]
    );
    book.issue_count = issueCount.count;

    res.json(book);
  } catch (error) {
    console.error('Error fetching book:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Search books
router.get('/search', async (req, res) => {
  try {
    const q = req.query.q || '';
    const filter = req.query.filter || 'all';

    if (!q) {
      return res.json({ books: [] });
    }

    let sql = 'SELECT * FROM books WHERE ';
    let params = [];

    if (filter === 'title') {
      sql += 'title LIKE ?';
      params.push(`%${q}%`);
    } else if (filter === 'author') {
      sql += 'author LIKE ?';
      params.push(`%${q}%`);
    } else if (filter === 'subject') {
      sql += 'subject LIKE ?';
      params.push(`%${q}%`);
    } else {
      sql += '(title LIKE ? OR author LIKE ? OR isbn LIKE ? OR subject LIKE ?)';
      params.push(`%${q}%`, `%${q}%`, `%${q}%`, `%${q}%`);
    }

    sql += ' LIMIT 10';

    const books = await db.all(sql, params);
    res.json({ books });
  } catch (error) {
    console.error('Error searching books:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Create book (librarian only)
router.post('/', requireAuth, async (req, res) => {
  try {
    const { isbn, title, author, subject, publisher, publication_year, total_copies, shelf_location } = req.body;

    if (!isbn || !title || !author || !subject) {
      return res.status(400).json({ error: 'ISBN, title, author, and subject are required' });
    }

    const result = await db.run(
      `INSERT INTO books (isbn, title, author, subject, publisher, publication_year, total_copies, available_copies, shelf_location)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [isbn, title, author, subject, publisher || null, publication_year || null, total_copies || 1, total_copies || 1, shelf_location || null]
    );

    const book = await db.get('SELECT * FROM books WHERE id = ?', [result.id]);
    res.status(201).json(book);
  } catch (error) {
    if (error.message.includes('UNIQUE constraint')) {
      return res.status(400).json({ error: 'ISBN already exists' });
    }
    console.error('Error creating book:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Update book (librarian only)
router.put('/:id', requireAuth, async (req, res) => {
  try {
    const { isbn, title, author, subject, publisher, publication_year, total_copies, available_copies, shelf_location } = req.body;

    const existingBook = await db.get('SELECT * FROM books WHERE id = ?', [req.params.id]);
    if (!existingBook) {
      return res.status(404).json({ error: 'Book not found' });
    }

    // Calculate available copies adjustment
    const newTotalCopies = total_copies !== undefined ? total_copies : existingBook.total_copies;
    const currentIssued = existingBook.total_copies - existingBook.available_copies;
    const newAvailableCopies = Math.max(0, newTotalCopies - currentIssued);

    await db.run(
      `UPDATE books 
       SET isbn = ?, title = ?, author = ?, subject = ?, publisher = ?, publication_year = ?, 
           total_copies = ?, available_copies = ?, shelf_location = ?, updated_at = CURRENT_TIMESTAMP
       WHERE id = ?`,
      [
        isbn || existingBook.isbn,
        title || existingBook.title,
        author || existingBook.author,
        subject || existingBook.subject,
        publisher !== undefined ? publisher : existingBook.publisher,
        publication_year !== undefined ? publication_year : existingBook.publication_year,
        newTotalCopies,
        available_copies !== undefined ? available_copies : newAvailableCopies,
        shelf_location !== undefined ? shelf_location : existingBook.shelf_location,
        req.params.id
      ]
    );

    const book = await db.get('SELECT * FROM books WHERE id = ?', [req.params.id]);
    res.json(book);
  } catch (error) {
    console.error('Error updating book:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Delete book (librarian only)
router.delete('/:id', requireAuth, async (req, res) => {
  try {
    const book = await db.get('SELECT * FROM books WHERE id = ?', [req.params.id]);
    if (!book) {
      return res.status(404).json({ error: 'Book not found' });
    }

    // Check if book has active transactions
    const activeTransactions = await db.get(
      'SELECT COUNT(*) as count FROM transactions WHERE book_id = ? AND status IN ("issued", "overdue")',
      [req.params.id]
    );

    if (activeTransactions.count > 0) {
      return res.status(400).json({ error: 'Cannot delete book with active transactions' });
    }

    await db.run('DELETE FROM books WHERE id = ?', [req.params.id]);
    res.json({ message: 'Book deleted successfully' });
  } catch (error) {
    console.error('Error deleting book:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;
